File: List_1202/jiqing/445.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

 //the difference between this q and #2 is that in this question we should each digit from end to start
 //so we use 2 stacks to store the element in each list so that we can read from end to start, the other thoughts
 //is as same as #2
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        Deque<Integer> stk1 = new LinkedList<>();
        Deque<Integer> stk2 = new LinkedList<>();
        while(l1 != null) {
            stk1.push(l1.val);
            l1 = l1.next;
        }
        while(l2 != null) {
            stk2.push(l2.val);
            l2 = l2.next;
        }
        int sum = 0;
        ListNode ans = null;
        while(!stk1.isEmpty() || !stk2.isEmpty() || sum != 0) {
            int a = stk1.isEmpty() ? 0 : stk1.pop();
            int b = stk2.isEmpty() ? 0 : stk2.pop();
            sum += a + b;
            ListNode cur = new ListNode(sum % 10);
            cur.next = ans;
            ans = cur;
            sum /= 10;
        }
        return ans;
    }
}

File: List_1202/zzx/445.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# 对于需要逆序的listnode
# 首先考虑用栈存储listnoded的value 然后进行计算
#curnode=ListNode(cur) curnode.next=res res=curnode 逆向新建链表
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        stack1=[]
        stack2=[]
        while l1:
            stack1.append(l1.val)
            l1=l1.next
        while l2:
            stack2.append(l2.val)
            l2=l2.next
        carry=0
        res=None
        while stack1 or stack2 or carry!=0:
            a=stack1.pop()if stack1 else 0
            b=stack2.pop() if stack2 else 0
            cur=a+b+carry
            carry=cur//10
            cur=cur%10
            curnode=ListNode(cur)
            curnode.next=res
            res=curnode
        return res
        
        